### PR TITLE
Decouple petra warp efficiency from speed up multiplier

### PR DESCRIPTION
**Pull request by sextondb**
The basic issue is that 1 warp is not equal to 1 second, contrary to the documentation and user expectation. If a user has 3600 warp due to 1 hour of offline time, then the expectation from the user perspective is that they will get 3600 "seconds" worth of production regardless of speed.

Unfortunately this is not the case. At +1x speed the user gets 1800 "seconds" of additional production for 3600 warp. Compare this to +7x speed where the user gets 3150 "seconds" of additional production for that same 3600 warp. 

Thus the current implementation heavily favors running Petra at 8x speed.

This PR changes the formula for Petra calculations to what is expected by the user. Each +1x of speedup, over the base, costs 1 warp per second. +0x costs 0 warp/second, +1x costs 1 warp/second, ..., +7x costs 7 warp/second.

This was accomplished by "smearing" the cost in warp/second so that it is smoothly paid over each tick. It was also done without cutting into the tick timer, in order to keep the complexity of the change low.

### CODE REVIEW

**Comment by sextondb** [comment]
I have seen Fish Pond's previous response on this topic, and I will attempt to address them here (in the order I found them in discord):


> I appreciate the suggestions. But the core issue here is that the game's logic should not differ due to the speed it's running.

The game's logic already runs differently based on the speed. If the speedup is 1, no Petra resources are used. If the speedup is1, then Petra resources are used. Additionally, the ratio between warp used and production received is changed when the speed is changed.

> The rule of 1 game cycle = 1 warp is required to be maintained.

The stated requirement is that 1 game cycle = 1 warp. However, the method of generation of warp itself encodes another equally important requirements. Warp itself is gained at a rate of "offline time in seconds". Which means that the requirement that "1 warp = 1 second" is required to be maintained as well. Petra already knows about it's own speed up. The 1 warp = 1 game cycle = 1 second is already encoded into the game engine. Indeed, Petra already knows this because it subtracts 1 warp per tick. If the game engine was suddenly changed to 60 (or 2 or whatever) ticks per second, then a change to Petra logic would already be necessary, either in the Petra code directly, or in the offline calculation code. In order for Petra to exist, something Petra related needs to know about the conversion rate between ticks, wall clock seconds, and warp.

> This is not a bug and I have explained it before 🙂

While it may not be a bug, it appears that it's not intended either. The fact that running it faster makes it more efficient is a side effect of the current code. It seems unlikely that the intent of Petra was to have it be more efficient at higher multipliers, and then follow that up with a warning that being more efficient is risky. Regardless, it can be considered a feature request.

> I agree it's not really intuitive, but this is what we can do without breaking the game's simulation

Petra already broke the simulation to some degree. It directly ties game ticks to seconds. Its code encodes "1 tick = 1 second". It just doesn't use a variable, but instead hides it via using the constant '1'. But that 1 is really just "warp cost per tick per second" in sheep's clothing.

>I know this can be confusing. But  Petra is always measured against game time, not wall clock time.

That's how it is currently coded. The feature request is to make "1 warp = 1 second". By definition, this is a request to change the feature to be more intuitive.

> When speeding up, 1 Time Warp = 1 game second. If you do 8x speed, you will still use 1 warp per game second, which is 8 warp per wall clock second.

The feature request is to align Warp with user's perceived definition of what they are getting for "offline time". This differs, as expected for a feature request, from the current implementation.

> I understand the your point. But if I implement it as you described, Petra would need to skip 1/2 cycles when running at 2x speed and 1/8 cycles when running at 8x speed, which is not supported by the logic.

The PR does not skip cycles.

> Because the speed of running the simulation should *not* change the resource consumption.

Again, this comes back to the feature request of "1 warp = 1 second". In order to create that relationship, a change in resource consumption is actually required.

tl;dr
Petra should give "1 second of production" for every "1 second of offline time". This matches user's expectations. This change couples Petra to the ratio of "game ticks per second", but Petra is already tied to the ratio of "game ticks per second" in a non-intuitive way.

**Comment by sextondb** [comment]
In the PR, Petra (mostly) only knows what Petra knows about the simulation. It knows it's own warp resource amount, and it knows the speed up it is to provide.

From a simulation perspective, the only thing that has changed is that the "Cost per tick per second" has been given a name instead of the constant value of 1 that the current implementation uses, and the logic has been then fixed to calculate the effective warp cost based on the "Cost per tick per second" value.
